### PR TITLE
[NEW] Enable the IDP to choose the best authnContext

### PR DESCRIPTION
**Pull request by felipecrp**
I would like to let the IDP choose the best authnContext for the request, but Rocket.Chat always send an authnContext. 

According to SAML specification, this information is optional in the request. So, I inserted an "if" to suppress this information when no authnContext is provided.

I still need help to test this patch.

Regards,

### CODE REVIEW

File: app/meteor-accounts-saml/server/saml_utils.js [hunk 1]
```diff
@@ -98,12 +98,14 @@ SAML.prototype.generateAuthorizeRequest = function(req) {
...
-		+= `<samlp:RequestedAuthnContext xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Comparison="${ authnContextComparison }">`
-		+ `<saml:AuthnContextClassRef xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">${ authnContext }</saml:AuthnContextClassRef></samlp:RequestedAuthnContext>\n`
-		+ '</samlp:AuthnRequest>';
+	if (this.options.authnContextComparison) {
+		const authnContextComparison = this.options.authnContextComparison || 'exact';
+		const authnContext = this.options.customAuthnContext || 'urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport';
+		request
+			+= `<samlp:RequestedAuthnContext xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Comparison="${ authnContextComparison }">`
+			+ `<saml:AuthnContextClassRef xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">${ authnContext }</saml:AuthnContextClassRef></samlp:RequestedAuthnContext>\n`
+			+ '</samlp:AuthnRequest>';
```
Comment (pierre-lehnen-rc) [inline app/meteor-accounts-saml/server/saml_utils.js:107]: You need to close the <samlp:AuthnRequest> tag outside the if block.

File: app/meteor-accounts-saml/server/saml_utils.js [hunk 2]
```diff
@@ -98,12 +98,14 @@ SAML.prototype.generateAuthorizeRequest = function(req) {
 		request += `<samlp:NameIDPolicy xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Format="${ this.options.identifierFormat }" AllowCreate="true"></samlp:NameIDPolicy>\n`;
 	}
 
-	const authnContextComparison = this.options.authnContextComparison || 'exact';
-	const authnContext = this.options.customAuthnContext || 'urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport';
-	request
-		+= `<samlp:RequestedAuthnContext xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" Comparison="${ authnContextComparison }">`
-		+ `<saml:AuthnContextClassRef xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">${ authnContext }</saml:AuthnContextClassRef></samlp:RequestedAuthnContext>\n`
-		+ '</samlp:AuthnRequest>';
+	if (this.options.authnContextComparison) {
```
Comment (pierre-lehnen-rc) [inline app/meteor-accounts-saml/server/saml_utils.js:101]: The setting uses a select field with multiple choices, but it doesn't have an option to leave it blank. With the current setting there is no way to enable this behavior.

File: app/meteor-accounts-saml/server/saml_utils.js (hunk 1 above)
Comment (felipecrp) [reply app/meteor-accounts-saml/server/saml_utils.js]: Thank you for the help. I changed the check condition to use customAuthnContext, which is a textbox.

File: app/meteor-accounts-saml/server/saml_utils.js (hunk 2 above)
Comment (felipecrp) [reply app/meteor-accounts-saml/server/saml_utils.js]: Fixed.

Comment (felipecrp) [reply app/meteor-accounts-saml/server/saml_utils.js]: Thank you for the help. I changed the check condition to use customAuthnContext, which is a textbox.

**Pull request merged**